### PR TITLE
⭐️ shodan provider

### PR DESCRIPTION
**Pull request by chris-rock**
# Shodan Provider

```shell
cnquery shell shodan
```

For authentication, you can use the `SHODAN_TOKEN` environment variable.

```shell
export SHODAN_TOKEN="<token>"
```

## Examples

**Host information**

Query the base information for a host by IP address.

```shell
cnquery> shodan.host("8.8.8.8") { * }
shodan.host: {
  tags: []
  hostnames: [
    0: "dns.google"
  ]
  org: "Google LLC"
  asn: "AS15169"
  ip: "8.8.8.8"
  isp: "Google LLC"
  vulnerabilities: null
  os: null
  ports: [
    0: 443
    1: 53
  ]
}
```

**Display the hostname for a host**

Query the hostname for a host.

```shell
cnquery> shodan.host("8.8.8.8").hostnames
shodan.host.hostnames: [
  0: "dns.google"
]
```

**Open ports for a host**

Display all open ports for a host.

```shell
cnquery> shodan.host("8.8.8.8").ports
shodan.host.ports: [
  0: 443
  1: 53
]
```

**DNS Lookup**

Query the DNS information for a domain.

```shell
cnquery> shodan.domain("example.com") { * }
shodan.domain: {
  name: "example.com"
  nsrecords: [
    0: shodan.nsrecord domain="example.com" subdomain="" type="A"
    1: shodan.nsrecord domain="example.com" subdomain="" type="AAAA"
    2: shodan.nsrecord domain="example.com" subdomain="" type="MX"
    3: shodan.nsrecord domain="example.com" subdomain="" type="NS"
    4: shodan.nsrecord domain="example.com" subdomain="" type="NS"
    5: shodan.nsrecord domain="example.com" subdomain="" type="SOA"
    6: shodan.nsrecord domain="example.com" subdomain="" type="TXT"
    7: shodan.nsrecord domain="example.com" subdomain="" type="TXT"
    8: shodan.nsrecord domain="example.com" subdomain="www" type="A"
    9: shodan.nsrecord domain="example.com" subdomain="www" type="AAAA"
    10: shodan.nsrecord domain="example.com" subdomain="www" type="TXT"
    11: shodan.nsrecord domain="example.com" subdomain="www" type="TXT"
  ]
  tags: [
    0: "ipv6"
    1: "spf"
  ]
  subdomains: [
    0: "www"
  ]
}
```

**Get all DNS NS records**

Query the DNS NS records for a domain.

```shell
cnquery> shodan.domain("example.com").nsrecords.where(type == "NS") { subdomain  type value }
shodan.domain.nsrecords.where: [
  0: {
    type: "NS"
    subdomain: ""
    value: "a.iana-servers.net"
  }
  1: {
    type: "NS"
    subdomain: ""
    value: "b.iana-servers.net"
  }
]
```

**Find all AAAA records for a subdomain**

Query the DNS AAAA records for  the "www" subdomain.

```shell
cnquery> shodan.domain("example.com").nsrecords.where(type == "AAAA").where(subdomain == "www") { subdomain  type value }
shodan.domain.nsrecords.where.where: [
  0: {
    subdomain: "www"
    value: "2606:2800:21f:cb07:6820:80da:af6b:8b2c"
    type: "AAAA"
  }
]
```

# Advanced Usage

Discover all exposed hosts on a network.

```shell
cnquery shell shodoan --networks "192.168.0.0/20" --discover hosts
```

Connect to a specific IP address and display all open ports.

```shell
cnquery shell shodan host 8.8.8.8
```

Connect to a domain and display subdomains.

```shell
cnquery shell shodan domain example.com
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -528,6 +538,11 @@ lr/docs/markdown: providers/lr
 		--description "The Operating Systems (OS) resource pack lets you use MQL to query and assess the security of your operating system packages and configuration." \
 		--docs-file providers/os/resources/os.lr.manifest.yaml \
 		--output ../docs/docs/mql/resources/os-pack
+	./lr markdown providers/shodan/resources/shodan.lr \
+		--pack-name "Shodan" \
+		--description "The Shodan resource pack lets you use MQL to query and assess the security of your Shodan service." \
```
Comment (tas50) [inline Makefile]: We probably want to update this to describe what data you're getting from Shodan in these resources

File: providers/github/config/config.go [hunk 2]
```diff
@@ -6,6 +6,7 @@ package config
 import (
 	"go.mondoo.com/cnquery/v11/providers-sdk/v1/inventory"
 	"go.mondoo.com/cnquery/v11/providers-sdk/v1/plugin"
+	"go.mondoo.com/cnquery/v11/providers/github/connection"
```
Comment (tas50) [inline providers/github/config/config.go]: Should this be in this PR?

File: Makefile [hunk 3]
```diff
@@ -528,6 +538,11 @@ lr/docs/markdown: providers/lr
...
+		--docs-file providers/slack/resources/shodan.lr.manifest.yaml \
```
Comment (tas50) [inline Makefile]:
```suggestion
		--docs-file providers/shodan/resources/shodan.lr.manifest.yaml \
```

File: providers/shodan/resources/shodan.lr [hunk 4]
```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Mondoo, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+option provider = "go.mondoo.com/cnquery/v11/providers/shodan"
+option go_package = "go.mondoo.com/cnquery/v11/providers/shodan/resources"
+
+// Shodan
```
Comment (tas50) [inline providers/shodan/resources/shodan.lr]:
```suggestion
// Shodan Search Engine
```

File: providers/shodan/resources/shodan.lr [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+shodan {}
+
+// Shodan Host Information
```
Comment (tas50) [inline providers/shodan/resources/shodan.lr]:
```suggestion
// Shodan Search Engine host information
```

File: providers/shodan/resources/shodan.lr [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  tags() []string
+  // List of hostnames assigned to the IP
+  hostnames() []string
+  // Open ports
+  ports() []int
+  // Vulnerabilities
+  vulnerabilities() []string
+}
+
+// Shodan Domain Information
```
Comment (tas50) [inline providers/shodan/resources/shodan.lr]:
```suggestion
// Shodan Search Engine domain information
```

File: providers/shodan/resources/shodan.lr [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  name string
+  // Domain tags
+  tags() []string
+  // List of subdomains
+  subdomains() []string
+  // DNS NS records
+  nsrecords() []shodan.nsrecord
+}
+
+// Shodan DNS NS Record
```
Comment (tas50) [inline providers/shodan/resources/shodan.lr]:
```suggestion
// Shodan Search Engine DNS NS record
```

File: providers/shodan/resources/shodan.lr [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  subdomain string
+  // DNS record types
+  type string
+  // DNS record value
+  value string
+  // Last seen by Shodan
+  lastSeen time
+}
+
+// Shodan account
```
Comment (tas50) [inline providers/shodan/resources/shodan.lr]:
```suggestion
// Shodan Search Engine account
```

File: providers/shodan/resources/shodan.lr [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+shodan.profile @defaults("displayName"){
+  member bool
+  credits int
+  displayName string
+  createdAt time
+}
+
+// Shodan API Plan Information
```
Comment (tas50) [inline providers/shodan/resources/shodan.lr]:
```suggestion
// Shodan Search Engine API plan information
```

File: providers/shodan/resources/shodan.lr.manifest.yaml [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  shodan.host:
+    fields:
+      asn: {}
+      hostnames: {}
+      ip: {}
+      isp: {}
+      org: {}
+      os: {}
+      ports:
+        min_mondoo_version: latest
```
Comment (tas50) [inline providers/shodan/resources/shodan.lr.manifest.yaml]: nit: these can get collapsed down to just {} right?

File: providers/github/config/config.go (hunk 2 above)
Comment (chris-rock) [reply providers/github/config/config.go]: We can extract it.

File: providers/shodan/resources/shodan.lr.manifest.yaml (hunk 10 above)
Comment (chris-rock) [reply providers/shodan/resources/shodan.lr.manifest.yaml]: +1

File: providers/shodan/config/config.go [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+					Long:    "token",
+					Type:    plugin.FlagType_String,
+					Default: "",
+					Desc:    "Shodan API token",
+				},
+				{
+					Long:    "networks",
+					Type:    plugin.FlagType_List,
+					Default: "",
+					Desc:    "Only include repositories with matching names.",
```
Comment (tas50) [inline providers/shodan/config/config.go]:
```suggestion
					Desc:    "Only include repositories with matching names",
```

**Comment by chris-rock** [comment]
@tas50 Thank you for the great feedback. PR adjusted.

**Comment by tas50** [comment]
Can you add this to the list in the main readme.md as well please?

**Pull request merged**